runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -3269,31 +3227,174 @@ MM_CopyForwardScheme::completeScanForAbort(MM_EnvironmentVLHGC *env)
 	} while (handleOverflow(env));
 }
 
+void
+MM_CopyForwardScheme::completeScanWorkPacket(MM_EnvironmentVLHGC *env)
+{
+	MM_AllocationContextTarok *reservingContext = _commonContext;
+	J9Object *objectPtr = NULL;
+
+	while (NULL != (objectPtr = (J9Object *)env->_workStack.popNoWaitFromCurrentInputPacket(env))) {
+		Assert_MM_false(MM_ScavengerForwardedHeader(objectPtr).isForwardedPointer());
+		scanObject(env, reservingContext, objectPtr, SCAN_REASON_PACKET);
+	}
+}
+
+MM_CopyForwardScheme::CopyForwardSchemeScanReason
+MM_CopyForwardScheme::getNextWorkUnitNoWait(MM_EnvironmentVLHGC *env, UDATA preferredNumaNode)
+{
+	UDATA nodeLists = _scanCacheListSize;
+	CopyForwardSchemeScanReason ret = SCAN_REASON_NONE;
+	/* local node first */
+	ret = getNextWorkUnitOnNode(env, preferredNumaNode);
+	if (SCAN_REASON_NONE == ret) {
+		/* we failed to find a scan cache on our preferred node */
+		if (COMMON_CONTEXT_INDEX != preferredNumaNode) {
+			/* try the common node */
+			ret = getNextWorkUnitOnNode(env, COMMON_CONTEXT_INDEX);
+		}
+		/* now try the remaining nodes */
+		UDATA nextNode = (preferredNumaNode + 1) % nodeLists;
+		while ((SCAN_REASON_NONE == ret) && (nextNode != preferredNumaNode)) {
+			if (COMMON_CONTEXT_INDEX != nextNode) {
+				ret = getNextWorkUnitOnNode(env, nextNode);
+			}
+			nextNode = (nextNode + 1) % nodeLists;
+		}
+	}
+	if (SCAN_REASON_NONE == ret && (0 != _regionCountCannotBeEvacuated) && !_abortInProgress && !abortFlagRaised()) {
+		if (env->_workStack.retrieveInputPacket(env)) {
+			ret = SCAN_REASON_PACKET;
+		}
+	}
+	return ret;
+}
+
+
+MM_CopyForwardScheme::CopyForwardSchemeScanReason
+MM_CopyForwardScheme::getNextWorkUnit(MM_EnvironmentVLHGC *env, UDATA preferredNumaNode)
```

Comment:
try to move this up closer to where getNextScanCache was, to help diff tool

Reply:
done